engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/SortNode.java

```diff
@@ -0,0 +1,121 @@
+import com.oracle.truffle.api.nodes.LoopNode;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.profiles.BranchProfile;
+import java.util.Arrays;
+import java.util.Comparator;
+import org.enso.interpreter.Language;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.callable.atom.Atom;
+import org.enso.interpreter.runtime.callable.atom.AtomConstructor;
+import org.enso.interpreter.runtime.callable.function.Function;
+import org.enso.interpreter.runtime.data.Array;
+import org.enso.interpreter.runtime.error.PanicException;
+import org.enso.interpreter.runtime.state.Stateful;
+import org.enso.interpreter.runtime.state.data.EmptyMap;
+
+@BuiltinMethod(type = "Array", name = "sort", description = "Sorts a mutable array in place.")
+public abstract class SortNode extends Node {
+  private @Child ComparatorNode comparatorNode = ComparatorNode.build();
+
+  abstract Object execute(VirtualFrame frame, Object _this, Object comparator);
+
+  static SortNode build() {
+    return SortNodeGen.create();
+  }
+
+  @Specialization
+  Object doSortFunction(
+      VirtualFrame frame,
+      Array _this,
+      Function comparator,
+      @CachedContext(Language.class) ContextReference<Context> ctxRef) {
+    Comparator<Object> compare = getComparator(comparator, ctxRef);
+    return runSort(compare, _this, ctxRef);
+  }
+
+  @Specialization
+  Object doSortCallable(
+      VirtualFrame frame,
+      Array _this,
+      Object comparator,
+      @CachedContext(Language.class) ContextReference<Context> ctxRef) {
+    Comparator<Object> compare = (l, r) -> comparatorNode.execute(frame, comparator, l, r);
+    return runSort(compare, _this, ctxRef);
+  }
+
+  @Specialization
+  Object doAtomThis(
+      VirtualFrame frame,
+      Atom _this,
+      Object that,
+      @CachedContext(Language.class) ContextReference<Context> ctxRef,
+      @Cached("ctxRef.get().getBuiltins().mutable().array()") AtomConstructor array) {
+    return ctxRef.get().getBuiltins().nothing().newInstance();
+  }
+
+  Object runSort(Comparator<Object> compare, Array _this, ContextReference<Context> ctxRef) {
+    doSort(_this.getItems(), compare);
+    LoopNode.reportLoopCount(this, _this.length());
+    return ctxRef.get().getBuiltins().nothing().newInstance();
+  }
+
+  @TruffleBoundary
+  void doSort(Object[] items, Comparator<Object> compare) {
+    Arrays.sort(items, compare);
+  }
+
+  private SortComparator getComparator(Function comp, ContextReference<Context> ctxRef) {
+    return new SortComparator(comp, ctxRef);
+  }
+
+  private class SortComparator implements Comparator<Object> {
+    private final Function compFn;
+    private final ContextReference<Context> ctxRef;
+    private final Atom less;
+    private final Atom equal;
+    private final Atom greater;
+    private final BranchProfile resultProfile = BranchProfile.create();
```

Comment:
This BranchProfile needs to belong to the node.

Reply:
See `usage` here https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/profiles/BranchProfile.html